File: snakemake/workflow/script/gtf2anno_tab.pl
use strict; use warnings;

print "gtf_2_anno_tab.gene_level.pl <xxx.gtf>\n";
print "Out: anno.txt\n";
print "good for Equus_caballus.EquCab3.0.104.gtf\n";

die unless @ARGV == 1;

open(GTF, shift) or die;

my%id2name;
while(<GTF>){
    chomp;
    my@e = split "\t", $_;
    next if /^#/;
    next if $e[2] ne 'gene';
    next if $_ !~ m/gene_id/;

    my($id) = $_ =~ /gene_id ([^;]+);/;
    my($name);

    if ($_ !~ m/gene_name/){
        $name = $id
    }else{
        ($name) = $_ =~ /gene_name ([^;]+);/;
    }

    my($type) = $_ =~ /gene_biotype ([^;]+);/;  # change: now for mouse gencode mm10

    if (length($type) == 0){$type = "-"}

    $name =~ s/^\s+//;
    $id =~ s/^\s+//;
    $type =~ s/^\s+//;

    $id2name{$id} = "$name\t$type";
}
close GTF;

open(OUT, ">anno.txt");
print OUT "Gene\tName\tType\n";
for (sort keys %id2name){
    my$out = "$_\t$id2name{$_}\n";
    $out =~ s/\"//g;
    print OUT $out;
}
close OUT;

File: snakemake/workflow/script/gtf2anno_tab.py
import sys
import re

file = sys.argv[1]
outfile = file + '.txt'
print("Input: {}\nOutput: {}".format(file,outfile))


o = open (outfile,'w')
o.write("\t".join(['Gene','Name','Type','Chr','Strand','Start','End'])+"\n")

with open(file) as f:
    while True:
        line = f.readline()
        if not line:
            break
        
        x = line.split("\t")

        if len(x) != 9:
            continue

        seqname,source,feature,start,end,score,strand,fname,attribute=x

        if feature != "gene":
            continue
    
        gene_id = re.search(r'gene_id "([^\"]+)"', attribute).groups()[0]
        if re.search(r'gene_name "([^\"]+)"', attribute): 
            gene_name = re.search(r'gene_name "([^\"]+)"', attribute).groups()[0]
        else:
            gene_name = gene_id
        gene_type = re.search(r'gene_biotype "([^\"]+)"', attribute).groups()[0]

        outline = "\t".join([gene_id,gene_name,gene_type,seqname,strand,start,end])+"\n"
        o.write(outline)

o.close()

